File: aramisbaselibrary/src/main/java/com/aramis/library/component/service/VersionService.java
package com.aramis.library.component.service;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.os.Bundle;
import android.os.IBinder;
import android.support.annotation.Nullable;
import android.support.v4.app.NotificationCompat;
import android.text.TextUtils;

import com.kymjs.rxvolley.client.FileRequest;
import com.kymjs.rxvolley.client.HttpCallback;
import com.kymjs.rxvolley.client.ProgressListener;
import com.kymjs.rxvolley.client.RequestConfig;

import java.io.File;

/**
 * VersionService 版本更新
 * Created by Aramis on 2017/6/25.
 */

public class VersionService {
//        extends Service {
//    private VersionBean versionBean;
//
//    @Nullable
//    @Override
//    public IBinder onBind(Intent intent) {
//        return null;
//    }
//
//    @Override
//    public void onCreate() {
//        super.onCreate();
//    }
//
//    @Override
//    public int onStartCommand(Intent intent, int flags, int startId) {
//        if (mNotificationManager == null) {
//            initNotification();
//        }
//        String url = "";
//        if (intent != null) {
//            versionBean = (VersionBean) intent.getSerializableExtra("bean");
//            url = versionBean.getVersionsUrl();
//        }
//        LogUtils.e("onStartCommand", "url:" + url);
//        if (versionBean != null && !TextUtils.isEmpty(url)) {
//            downloadFile(url);
//        }
//        return super.onStartCommand(intent, flags, startId);
//    }
//
//    private long cc = 0;
//
//    private String getDownloadFileFolder() {
//        return GlobalPrams.getRootFolder() + File.separator + "apk";
//    }
//
//    private String getDownloadFilePath() {
//        if (versionBean != null) {
//            return getDownloadFileFolder() + File.separator + "yhh" + versionBean.getVersionsName() + ".apk";
//        }
//        return "";
//    }
//
//    private void downloadFile(String url) {
//        File file = new File(getDownloadFileFolder());
//        if (!file.exists()) file.mkdirs();
////        File file1 = new File(GlobalPrams.getApkFilePath());
////        if (file1.exists()) file1.delete();
//        //下载文件路径
//        String filePath = getDownloadFilePath();
//
//        RequestConfig requestConfig = new RequestConfig();
//        requestConfig.mUrl = url;
//        requestConfig.mRetryPolicy = new SingalRetryPolicy();
//        requestConfig.mTimeout = 90 * 1000;
//        FileRequest fileRequest = new FileRequest(filePath, requestConfig,
//                new HttpCallback() {
//                    @Override
//                    public void onSuccess(String t) {
//                        super.onSuccess(t);
//                        LogUtils.e("onSuccess", "下载成功");
//                        ArBus.getDefault().post(getBundle(true, 100, 100));
//                        notifyNotification("下载完成", 100);
//                        ApkUtils.installApk(VersionService.this, getDownloadFilePath());
//                        setNotificationClick();
//                        stopSelf();
//                    }
//
//                    @Override
//                    public void onFailure(int errorNo, String strMsg) {
//                        super.onFailure(errorNo, strMsg);
//                        LogUtils.e("onSuccess", "下载失败");
//                        Bundle bundle = new Bundle();
//                        bundle.putBoolean("isUpdateFail", true);
//                        notifyNotification("下载失败", 0);
//                        ArBus.getDefault().post(bundle);
//                        stopSelf();
//                    }
//                });
//        fileRequest.setOnProgressListener(new ProgressListener() {
//            @Override
//            public void onProgress(long transferredBytes, long totalSize) {
////                LogUtils.e("开始进度", "per:" + transferredBytes + ",total:" + totalSize);
//                long l = System.currentTimeMillis();
//                if (l - cc >= 1000) {
//                    LogUtils.e("开始进度", "per:" + transferredBytes + ",total:" + totalSize);
//                    cc = l;
//                    notifyNotification("下载中", (int) (transferredBytes * 100 / totalSize));
//                    ArBus.getDefault().post(getBundle(false, transferredBytes, totalSize));
//                }
//            }
//        });
//        new ArRxVolley.Builder().setRequest(fileRequest).doTask();
//    }
//
//    private Bundle getBundle(boolean isFinish, long transferredBytes, long totalSize) {
//        Bundle bundle = new Bundle();
//        bundle.putBoolean("isUpdateVersion", true);
//        bundle.putBoolean("isUpdateFinish", isFinish);
//        bundle.putLong("total", totalSize);
//        bundle.putLong("transfer", transferredBytes);
//        if (totalSize != 0) {
//            bundle.putDouble("percent", transferredBytes / totalSize);
//        }
//        return bundle;
//    }
//
//    private NotificationManager mNotificationManager;
//    private NotificationCompat.Builder mBuilder;
//
//    /**
//     * 更新通知
//     */
//    private void notifyNotification(String title, int percent) {
//        if (!TextUtils.isEmpty(title)) {
//            mBuilder.setContentTitle(title);
//        }
//        mBuilder.setContentText(percent + "%").setProgress(100, percent, false);
//        mNotificationManager.notify(12, mBuilder.build());
//    }
//
//    private void initNotification() {
//        mNotificationManager =
//                (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
//        mBuilder = new NotificationCompat.Builder(this);
//        mBuilder.setContentTitle("正在下载")
//                .setContentText("0%")
//                .setSmallIcon(R.mipmap.ic_launcher);
//        mBuilder.setAutoCancel(true);
//    }
//
//    private void setNotificationClick() {
//        Intent notifyIntent = new Intent(this, OldMainActivity.class);
//        notifyIntent.putExtra("apkPath", getDownloadFilePath());
////        notifyIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
//        PendingIntent pendingNotifyIntent =
//                PendingIntent.getActivity(
//                        this,
//                        0,
//                        notifyIntent,
//                        PendingIntent.FLAG_UPDATE_CURRENT
//                );
//
//        mBuilder.setContentIntent(pendingNotifyIntent);
//    }
}

File: app/src/main/java/com/moviegetter/service/VersionService.kt
package com.moviegetter.service

import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Intent
import android.os.Bundle
import android.os.IBinder
import android.support.v4.app.NotificationCompat
import com.aramis.library.aramis.ArBus
import com.aramis.library.extentions.logE
import com.aramis.library.http.ArRxVolley
import com.aramis.library.http.SingalRetryPolicy
import com.aramis.library.utils.ApkUtils
import com.kymjs.rxvolley.client.FileRequest
import com.kymjs.rxvolley.client.HttpCallback
import com.kymjs.rxvolley.client.RequestConfig
import com.kymjs.rxvolley.http.VolleyError
import com.moviegetter.R
import com.moviegetter.ui.MainActivity
import org.jetbrains.anko.toast
import java.io.File

/**
 *Created by Aramis
 *Date:2018/9/17
 *Description:
 */
class VersionService : Service() {
    private var mNotificationManager: NotificationManager? = null
    private var mBuilder: NotificationCompat.Builder? = null

    private var downloadUrl = ""
    private var downFilePath = ""
    private var downloadFileFolder = ""
    private var lastL = 0L

    override fun onBind(intent: Intent?): IBinder? {
        return null
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {

        if (mNotificationManager == null) {
            initNotification()
        }
        intent?.apply {
            downloadUrl = intent.getStringExtra("downloadUrl")
            downFilePath = intent.getStringExtra("downFilePath")

            logE("downloadUrl:$downloadUrl,downFilePath:$downFilePath")

            if (downloadUrl.isNotBlank() && downFilePath.isNotBlank()) {
                downloadFileFolder = downFilePath.substring(0, downFilePath.lastIndexOf("/"))
                downloadFile(downloadUrl)
            } else {
                toast("params null. downloadUrl:$downloadUrl,downFilePath:$downFilePath")
            }
        }
        return super.onStartCommand(intent, flags, startId)
    }

    private fun downloadFile(url: String) {
        val file = File(downloadFileFolder);
        if (!file.exists()) file.mkdirs();
//        File file1 = new File(GlobalPrams.getApkFilePath());
//        if (file1.exists()) file1.delete();
        //下载文件路径
        val filePath = downFilePath

        val requestConfig = RequestConfig()
        requestConfig.mUrl = url
        requestConfig.mRetryPolicy = SingalRetryPolicy();
        requestConfig.mTimeout = 90 * 1000
        val fileRequest = FileRequest(filePath, requestConfig,
                object : HttpCallback() {

                    override fun onSuccess(t: String?) {
                        super.onSuccess(t)
                        ArBus.getDefault().post(getBundle(true, 100, 100));
                        notifyNotification("下载完成", 100);
                        ApkUtils.installApk(this@VersionService, downFilePath);
                        setNotificationClick();
                        stopSelf();
                    }

                    override fun onFailure(error: VolleyError?) {
                        super.onFailure(error)
                        logE("$error")
                        val bundle = Bundle()
                        bundle.putBoolean("isUpdateFail", true)
                        notifyNotification("下载失败", 0)
                        ArBus.getDefault().post(bundle)
                        stopSelf()
                    }

                })
        fileRequest.setOnProgressListener { transferredBytes, totalSize ->
            val l = System.currentTimeMillis();
            if (l - lastL >= 1000) {
//                    LogUtils.e("开始进度", "per:" + transferredBytes + ",total:" + totalSize);
                lastL = l
                notifyNotification("下载中", (transferredBytes * 100 / totalSize).toInt())
                ArBus.getDefault().post(getBundle(false, transferredBytes, totalSize))
            }
        }
        ArRxVolley.Builder().setRequest(fileRequest).doTask()
    }

    private fun notifyNotification(title: String, percent: Int) {
        if (title.isNotBlank()) {
            mBuilder?.setContentTitle(title);
        }
        mBuilder?.setContentText(percent.toString() + "%")?.setProgress(100, percent, false);
        mNotificationManager?.notify(12, mBuilder?.build())
    }

    private fun getBundle(isFinish: Boolean, transferredBytes: Long, totalSize: Long): Bundle {
        val bundle = Bundle();
        bundle.putBoolean("isUpdateVersion", true);
        bundle.putBoolean("isUpdateFinish", isFinish);
        bundle.putLong("total", totalSize);
        bundle.putLong("transfer", transferredBytes);
        if (totalSize != 0L) {
            bundle.putDouble("percent", transferredBytes.toDouble() / totalSize)
        }
        return bundle
    }

    private fun initNotification() {
        mNotificationManager = getSystemService(NOTIFICATION_SERVICE) as NotificationManager
        mBuilder = NotificationCompat.Builder(this);
        mBuilder?.setContentTitle("正在下载")?.setContentText("0%")?.setSmallIcon(R.mipmap.ic_launcher);
        mBuilder?.setAutoCancel(true);
    }

    private fun setNotificationClick() {
        val notifyIntent = Intent(this, MainActivity::class.java)
        notifyIntent.putExtra("apkPath", downFilePath);
//        notifyIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        val pendingNotifyIntent =
                PendingIntent.getActivity(
                        this,
                        0,
                        notifyIntent,
                        PendingIntent.FLAG_UPDATE_CURRENT
                );

        mBuilder?.setContentIntent(pendingNotifyIntent);
    }
}